docs/guides/advanced/authentication.md

```diff
@@ -4,7 +4,523 @@ title: Authentication
+
+<img src="../../../images/auth_endpoint.png" srcset="../../../images/auth_endpoint_2x.png 2x" class="screenshot" alt="Setting the endpoint URL">
+
+[View Sample Code][sample_manual_endpoint]{ .md-button }
+
+!!! warning "Not compatible with OAuth2"
+    Packs that use OAuth2 authentication don't support a manual prompt for the endpoint URL. They should instead use one of the alternate options listed below.
+
+
+### Extracting from token exchange
+
+Some APIs that use OAuth2 authentication return the endpoint URL during the token exchange, as a custom property within the JSON body:
+
+```json
+{
+  "access_token": "...",
+  "token_type": "bearer",
+  "site_url": "https://vandelay.example.com"
+}
+```
+
+You can automatically extract this value and use it as the endpoint URL by setting the `endpointKey` field of the authentication configuration.
+
+```ts
+pack.setUserAuthentication({
+  // ...
+  endpointKey: "site_url",
+});
+```
+
+[View Sample Code][sample_automatic_endpoint]{ .md-button }
+
+
+### Prompting for user selection
+
+If the service allows the same user account to access multiple endpoints, and those endpoints can be determined with an API call, then you can prompt the user to select the endpoint they wish to connect to. This is accomplished using a `postSetup` step of the type `SetEndpoint`.
+
+```ts
+pack.setUserAuthentication({
+  // After approving access, the user should select which instance they want to
+  // connect to.
+  postSetup: [{
+    type: coda.PostSetupType.SetEndpoint,
+    name: "SelectEndpoint",
+    description: "Select the site to connect to:",
+    // Generate the list of endpoint options.
+    getOptionsFormula: coda.makeMetadataFormula(async function (context) {
+      // Make a request to the API to retrieve the sites they can access.
+      let response = await context.fetcher.fetch({
+        method: "GET",
+        url: "https://api.example.com/my/sites",
+      });
+      let sites = response.body.sites;
+      return sites.map(site => {
+        return { display: site.name, value: site.url };
+      });
+    }),
+  }],
+});
+```
+
+The step's `getOptionsFormula` works like [dynamic autocomplete][autocomplete_dynamic], fetching the list of available endpoints and returning a display name and value for each. After the user enters their credentials they will be prompted to select one of the endpoints in a dialog.
+
+<img src="../../../images/auth_setendpoint.png" srcset="../../../images/auth_setendpoint_2x.png 2x" class="screenshot" alt="Selecting from a list of endpoints">
+
+[View Sample Code][sample_selected_endpoint]{ .md-button }
+
+!!! warning "Connection name generated twice"
+    When using this feature the [`getConnectionName` function](#setting-account-names) is run multiple times: once before the endpoint is selected and again after. Make sure your code works correctly in both cases. Before the endpoint is available you can return a generic name, which will get overwritten later.
```

Comment:
Ah interesting, I didn't realize this. Wonder if we can clean this up.

Reply:
That would be nice. I have a TODO to file a related bug, which is that if your `getOptionFormula` fails with an error, your partially complete account sticks around, in an unusable state (since it has no endpoint set).

Reply:
Bug filed: https://golinks.io/bug/20672